sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala

```diff
@@ -278,24 +278,20 @@ object JavaTypeInference {
 
       case _ if mapType.isAssignableFrom(typeToken) =>
         val (keyType, valueType) = mapKeyValueType(typeToken)
-        val keyDataType = inferDataType(keyType)._1
-        val valueDataType = inferDataType(valueType)._1
 
         val keyData =
           Invoke(
-            MapObjects(
+            UnresolvedMapObjects(
               p => deserializerFor(keyType, Some(p)),
-              Invoke(getPath, "keyArray", ArrayType(keyDataType)),
-              keyDataType),
+              UnresolvedGetArrayFromMap(getPath, GetArrayFromMap.Key())),
```

Comment:
Seems we don't need to make in unresolved

```
case class GetArrayFromMap(map: Expression, getKey: Boolean) extends Expression = {
  override def inputTypes = Seq(MapType)

  override def dataType = {
    val MapType(kt, vt) = map.dataType.asInstanceOf[MapType]
    if (getKey) kv else vt
  }

  override def eval...
 
  override def doCodegen...
}

```

Reply:
My idea was to replace this unresolved expression with _Invoke_.
Do you mean to write _eval_ and _doGenCode_ from scratch?
I don't see a way to wrap _Invoke_ as _doGenCode_ is protected and can't be invoked.

Reply:
Yea we can write eval and doGenCode from scratch. It's also more efficient since we can omit the useless try-catch in `Invoke`.

e.g.
```
// from UnaryExpression
override def nullSafeEval(input: Any) = input.asInstanceOf[MapData].keys

override def doGenCode = defineCodeGen(ctx, ev, c => s"$c.keys")
```

Reply:
I see, OK!